File: killrvideo-service-sugestedvideo/src/main/java/com/killrvideo/service/sugestedvideo/dao/SuggestedVideosMessagingDaoSupport.java
package com.killrvideo.service.sugestedvideo.dao;

import java.util.Date;
import java.util.UUID;

import com.killrvideo.service.sugestedvideo.grpc.SuggestedVideosServiceGrpcMapper;
import com.killrvideo.service.sugestedvideo.repository.SuggestedVideosRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.killrvideo.utils.GrpcMappingUtils;

import killrvideo.ratings.events.RatingsEvents.UserRatedVideo;
import killrvideo.user_management.events.UserManagementEvents.UserCreated;
import killrvideo.video_catalog.events.VideoCatalogEvents.YouTubeVideoAdded;

import javax.inject.Inject;

/**
 * Message processing for suggestion services.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
public abstract class SuggestedVideosMessagingDaoSupport {
    
    /** Loger for that class. */
    private static Logger LOGGER = LoggerFactory.getLogger(SuggestedVideosMessagingDaoSupport.class);
    
    @Inject
    protected SuggestedVideosRepository suggestedVideosRepository;
    @Inject
    private SuggestedVideosServiceGrpcMapper mapper;

    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     * 
     * @param userVideoRated
     *      a user has rated a video event
     */
    protected void onVideoRatingMessage(UserRatedVideo userVideoRated) {
        String videoId = userVideoRated.getVideoId().getValue();
        UUID   userId  = UUID.fromString(userVideoRated.getUserId().getValue());
        int rating     = userVideoRated.getRating();
        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug("[NewUserEvent] Processing rating with user {} and video {}", userId, videoId);
        }
        suggestedVideosRepository.updateGraphNewUserRating(videoId, userId, rating);
    }
    
    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     * 
     * @param userCreationMessage
     *      a user has been created
     */
    protected void onUserCreatingMessage(UserCreated userCreationMessage) {
        final UUID userId       = UUID.fromString(userCreationMessage.getUserId().getValue());
        final Date userCreation = GrpcMappingUtils.timestampToDate(userCreationMessage.getTimestamp());
        final String email      = userCreationMessage.getEmail();
        if (LOGGER.isDebugEnabled()) {
            LOGGER.debug("[NewUserEvent] Processing for user {} ", userId);
        }
        suggestedVideosRepository.updateGraphNewUser(userId, email, userCreation);
    }
    
    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     * 
     * @param videoAdded
     *      a video has been created
     */
    protected void onYoutubeVideoAddingMessage(YouTubeVideoAdded videoAdded) {
       suggestedVideosRepository.updateGraphNewVideo(mapper.mapVideoAddedtoVideoDTO(videoAdded));
    }
    

}

File: killrvideo-service-suggestedvideo/src/main/kotlin/com/killrvideo/service/suggestedvideo/dao/SuggestedVideosMessagingDaoSupport.kt
package com.killrvideo.service.suggestedvideo.dao

import com.killrvideo.service.suggestedvideo.grpc.SuggestedVideosServiceGrpcMapper
import com.killrvideo.service.suggestedvideo.repository.SuggestedVideosRepository
import com.killrvideo.utils.GrpcMappingUtils.fromUuid
import com.killrvideo.utils.GrpcMappingUtils.timestampToDate
import killrvideo.ratings.events.RatingsEvents.UserRatedVideo
import killrvideo.user_management.events.UserManagementEvents.UserCreated
import killrvideo.video_catalog.events.VideoCatalogEvents.YouTubeVideoAdded
import mu.KotlinLogging

/**
 * Message processing for suggestion services.
 *
 * @author Cedrick LUNVEN (@clunven)
 */
abstract class SuggestedVideosMessagingDaoSupport(
    private val suggestedVideosRepository: SuggestedVideosRepository,
    private val mapper: SuggestedVideosServiceGrpcMapper
) {
    private val logger = KotlinLogging.logger { }

    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     *
     * @param userVideoRated
     * a user has rated a video event
     */
    protected open fun onVideoRatingMessage(userVideoRated: UserRatedVideo) {
        val videoId = userVideoRated.videoId.value
        val userId = fromUuid(userVideoRated.userId)
        val rating = userVideoRated.rating
        logger.debug { "[NewUserEvent] Processing rating with user $userId and video $videoId" }
        suggestedVideosRepository.updateGraphNewUserRating(videoId, userId, rating)
    }

    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     *
     * @param userCreationMessage
     * a user has been created
     */
    protected open fun onUserCreatingMessage(userCreationMessage: UserCreated) {
        val userId = fromUuid(userCreationMessage.userId)
        val userCreation = timestampToDate(userCreationMessage.timestamp)
        val email = userCreationMessage.email
        logger.debug {"[NewUserEvent] Processing for user $userId " }
        suggestedVideosRepository.updateGraphNewUser(userId, email, userCreation)
    }

    /**
     * Message is consumed from specialized class but treatment is the same, updating graph.
     *
     * @param videoAdded
     * a video has been created
     */
    protected open fun onYoutubeVideoAddingMessage(videoAdded: YouTubeVideoAdded) {
        suggestedVideosRepository.updateGraphNewVideo(mapper.mapVideoAddedtoVideoDTO(videoAdded))
    }
}
